File: src/ts/src/local-file-recovery.ts
import { lstat, rmdir, unlink } from "node:fs/promises";
import { join } from "node:path";

import {
  LOCAL_FILE_LOCK_OWNER_FILE,
  LocalFileLockError,
  MAX_LOCAL_FILE_LOCK_OWNER_CODEPOINTS,
  MAX_LOCAL_FILE_LOCK_OWNER_UTF8_BYTES,
  read_bounded_local_file_lock_owner,
  read_local_file_lock_entry_names_bounded,
  validate_local_file_lock_owner,
  validate_local_file_lock_path,
} from "./local-file.js";

export { MAX_LOCAL_FILE_LOCK_OWNER_UTF8_BYTES } from "./local-file.js";

export type LocalFileLockInspectionState = "absent" | "held" | "incomplete" | "compromised";

export type LocalFileLockInspectionReason =
  | "owner_marker_missing"
  | "path_not_directory"
  | "dirty_directory"
  | "owner_not_regular_file"
  | "owner_too_large"
  | "owner_invalid_utf8"
  | "owner_identity_changed"
  | "permissions_widened"
  | "owner_contract_violation";

export type LocalFileLockInspection =
  | { state: "absent" }
  | { state: "held"; owner: string }
  | {
      state: "incomplete";
      reason: "owner_marker_missing";
      message: string;
    }
  | {
      state: "compromised";
      reason: LocalFileLockInspectionReason;
      message: string;
    };

type InspectionOwnerRead =
  | { state: "owner"; owner: string }
  | { state: "absent" }
  | { state: "incomplete" };

/** Read-only inspection. This never acquires, repairs, or removes a lock. */
export async function inspect_local_file_lock(path: string): Promise<LocalFileLockInspection> {
  validate_local_file_lock_path(path);
  const lockMetadata = await lstat_inspection_lock_path(path);
  if (lockMetadata === null) return { state: "absent" };
  if (!lockMetadata.isDirectory() || lockMetadata.isSymbolicLink()) {
    return compromised("path_not_directory", "lock path is not an unaliased directory");
  }
  if (process.platform !== "win32" && (lockMetadata.mode & 0o077) !== 0) {
    return compromised(
      "permissions_widened",
      "lock directory permissions widened beyond the private POSIX contract",
    );
  }

  const entries = await read_inspection_entry_names(path);
  if (entries === null) return { state: "absent" };
  if (entries.length === 0) {
    return incomplete(
      "lock directory has no owner marker; acquisition or release may have crashed mid-transition",
    );
  }
  if (entries.length !== 1 || entries[0] !== LOCAL_FILE_LOCK_OWNER_FILE) {
    return compromised("dirty_directory", "lock directory must contain exactly one owner marker");
  }

  const ownerPath = join(path, LOCAL_FILE_LOCK_OWNER_FILE);
  let ownerRead: InspectionOwnerRead;
  try {
    ownerRead = await read_inspection_owner(path, ownerPath);
  } catch (error) {
    if (error instanceof LocalFileLockError && error.kind === "compromised") {
      return compromised(classify_bounded_owner_error(error), error.message);
    }
    throw error;
  }
  if (ownerRead.state === "absent") return { state: "absent" };
  if (ownerRead.state === "incomplete") {
    return incomplete("owner token disappeared during inspection");
  }

  const owner = ownerRead.owner;
  if (owner.length === 0) {
    return compromised("owner_contract_violation", "owner token is empty");
  }
  if (Array.from(owner).length > MAX_LOCAL_FILE_LOCK_OWNER_CODEPOINTS) {
    return compromised(
      "owner_contract_violation",
      "owner token exceeds the portable 512-code-point contract bound",
    );
  }
  return { state: "held", owner };
}

/**
 * Explicit operator-driven recovery. `confirmed_inactive` asserts that the
 * caller independently established the previous owner is inactive and the
 * protected local state is quiescent. Missing locks are idempotent no-ops.
 * Ownerless incomplete crash-window state is intentionally never auto-recovered
 * because it has no owner identity to authenticate against.
 */
export async function recover_local_file_lock(
  path: string,
  expected_owner: string,
  confirmed_inactive: boolean,
): Promise<boolean> {
  validate_local_file_lock_path(path);
  validate_local_file_lock_owner(path, expected_owner);
  if (!confirmed_inactive) {
    throw new LocalFileLockError(
      "invalid_input",
      path,
      "explicit confirmed_inactive=true is required for recovery",
    );
  }

  const inspection = await inspect_local_file_lock(path);
  if (inspection.state === "absent") return false;
  if (inspection.state === "incomplete") {
    throw new LocalFileLockError(
      "compromised",
      path,
      "incomplete lock state has no owner identity; refusing automatic recovery",
    );
  }
  if (inspection.state === "compromised") {
    throw new LocalFileLockError("compromised", path, inspection.message);
  }
  if (inspection.owner !== expected_owner) {
    throw new LocalFileLockError(
      "compromised",
      path,
      "owner token does not match expected recovery owner",
    );
  }

  const finalInspection = await inspect_local_file_lock(path);
  if (finalInspection.state !== "held" || finalInspection.owner !== expected_owner) {
    throw new LocalFileLockError(
      "compromised",
      path,
      "local lock changed during recovery; refusing deletion",
    );
  }

  try {
    await unlink(join(path, LOCAL_FILE_LOCK_OWNER_FILE));
    await rmdir(path);
  } catch (error) {
    const code = error_code(error);
    throw new LocalFileLockError(
      code === "ENOTEMPTY" || code === "EEXIST" ? "compromised" : "io",
      path,
      `recover local lock failed: ${describe_error(error)}`,
      error,
    );
  }
  return true;
}

/**
 * Initial path metadata can itself observe Windows deletion-pending EPERM.
 * Retry only that transient condition, bounded to a few milliseconds. A path
 * that disappears linearizes as absent; persistent denial remains an IO error.
 */
async function lstat_inspection_lock_path(path: string) {
  const maxAttempts = process.platform === "win32" ? 4 : 1;
  let lastError: unknown;

  for (let attempt = 0; attempt < maxAttempts; attempt += 1) {
    try {
      return await lstat(path);
    } catch (error) {
      lastError = error;
      if (error_code(error) === "ENOENT") return null;
      if (process.platform !== "win32" || error_code(error) !== "EPERM") {
        throw io_error(path, "inspect local lock path", error);
      }
      if (attempt + 1 < maxAttempts) {
        await sleep_ms(1);
        continue;
      }
    }
  }

  throw io_error(path, "inspect local lock path", lastError);
}

/**
 * Read the bounded directory shape while tolerating only the two OS-level
 * disappearance signals produced by a concurrent clean release.
 *
 * POSIX typically reports ENOENT when rmdir wins between lstat and opendir.
 * Windows can transiently report EPERM while the directory is deletion-pending.
 * For Windows EPERM we perform a tiny bounded retry: disappearance linearizes
 * as `absent`; a persistent EPERM on an extant path remains a genuine IO error.
 */
async function read_inspection_entry_names(path: string): Promise<string[] | null> {
  const maxAttempts = process.platform === "win32" ? 4 : 1;
  let lastError: unknown;

  for (let attempt = 0; attempt < maxAttempts; attempt += 1) {
    try {
      return await read_local_file_lock_entry_names_bounded(path);
    } catch (error) {
      lastError = error;
      if (!(error instanceof LocalFileLockError) || error.kind !== "io") throw error;
      if (caused_by_error_code(error, "ENOENT")) return null;

      if (process.platform !== "win32" || !caused_by_error_code(error, "EPERM")) {
        throw error;
      }

      try {
        await lstat(path);
      } catch (probeError) {
        if (error_code(probeError) === "ENOENT") return null;
        if (error_code(probeError) !== "EPERM") {
          throw io_error(path, "recheck Windows deletion-pending local lock path", probeError);
        }
      }

      if (attempt + 1 < maxAttempts) {
        await sleep_ms(1);
        continue;
      }
    }
  }

  throw lastError;
}

/**
 * Read the owner marker for diagnostics while preserving fail-closed release
 * behavior. Windows may report EPERM when a clean concurrent release has
 * already put the owner/directory into deletion-pending state. Inspection may
 * linearize that bounded race as absent/incomplete, but persistent EPERM on an
 * extant lock remains an IO error and is never converted to healthy state.
 */
async function read_inspection_owner(
  path: string,
  ownerPath: string,
): Promise<InspectionOwnerRead> {
  const maxAttempts = process.platform === "win32" ? 4 : 1;
  let lastError: unknown;

  for (let attempt = 0; attempt < maxAttempts; attempt += 1) {
    try {
      return { state: "owner", owner: await read_bounded_local_file_lock_owner(path, ownerPath) };
    } catch (error) {
      lastError = error;

      if (error instanceof LocalFileLockError && error.kind === "compromised") {
        if (error.message.includes("owner token is missing")) {
          try {
            await lstat(path);
            return { state: "incomplete" };
          } catch (probeError) {
            if (error_code(probeError) === "ENOENT") return { state: "absent" };
            if (
              process.platform === "win32" &&
              error_code(probeError) === "EPERM" &&
              attempt + 1 < maxAttempts
            ) {
              await sleep_ms(1);
              continue;
            }
            throw io_error(path, "recheck local lock after owner disappearance", probeError);
          }
        }
        throw error;
      }

      if (
        !(error instanceof LocalFileLockError) ||
        error.kind !== "io" ||
        process.platform !== "win32" ||
        !caused_by_error_code(error, "EPERM")
      ) {
        throw error;
      }

      try {
        await lstat(path);
      } catch (probeError) {
        if (error_code(probeError) === "ENOENT") return { state: "absent" };
        if (error_code(probeError) !== "EPERM") {
          throw io_error(path, "recheck Windows deletion-pending local lock path", probeError);
        }
      }

      if (attempt + 1 < maxAttempts) {
        await sleep_ms(1);
        continue;
      }
    }
  }

  throw lastError;
}

function incomplete(message: string): LocalFileLockInspection {
  return { state: "incomplete", reason: "owner_marker_missing", message };
}

function compromised(
  reason: LocalFileLockInspectionReason,
  message: string,
): LocalFileLockInspection {
  return { state: "compromised", reason, message };
}

function classify_bounded_owner_error(error: LocalFileLockError): LocalFileLockInspectionReason {
  if (error.message.includes("valid UTF-8")) return "owner_invalid_utf8";
  if (error.message.includes("identity changed")) return "owner_identity_changed";
  if (error.message.includes("permissions widened")) return "permissions_widened";
  if (error.message.includes("multiple filesystem links") || error.message.includes("multiply linked")) {
    return "owner_identity_changed";
  }
  if (error.message.includes("exceeds the portable") && error.message.includes("byte")) {
    return "owner_too_large";
  }
  if (error.message.includes("regular file")) return "owner_not_regular_file";
  return "owner_contract_violation";
}

function sleep_ms(delay: number): Promise<void> {
  return new Promise((resolve) => setTimeout(resolve, delay));
}

function caused_by_error_code(error: Error, code: string): boolean {
  if (!("cause" in error)) return false;
  return error_code((error as Error & { cause?: unknown }).cause) === code;
}

function error_code(error: unknown): string | undefined {
  if (typeof error === "object" && error !== null && "code" in error) {
    const code = (error as { code?: unknown }).code;
    return typeof code === "string" ? code : undefined;
  }
  return undefined;
}

function describe_error(error: unknown): string {
  return error instanceof Error ? error.message : String(error);
}

function io_error(path: string, operation: string, error: unknown): LocalFileLockError {
  return new LocalFileLockError("io", path, `${operation} failed: ${describe_error(error)}`, error);
}

File: src/go/local_file_recovery.go
package oreslocks

import (
	"errors"
	"io"
	"os"
	"path/filepath"
	"runtime"
	"strings"
	"time"
	"unicode/utf8"
)

const localFileOwnerRecoveringName = "owner.recovering"

// LocalFileLockInspectionState describes read-only portable-lock state.
type LocalFileLockInspectionState string

const (
	LocalFileLockAbsent      LocalFileLockInspectionState = "absent"
	LocalFileLockHeld        LocalFileLockInspectionState = "held"
	LocalFileLockIncomplete  LocalFileLockInspectionState = "incomplete"
	LocalFileLockCompromised LocalFileLockInspectionState = "compromised"
)

// LocalFileLockInspectionReason is a stable machine-readable diagnostic code.
type LocalFileLockInspectionReason string

const (
	LocalFileOwnerMarkerMissing     LocalFileLockInspectionReason = "owner_marker_missing"
	LocalFilePathNotDirectory       LocalFileLockInspectionReason = "path_not_directory"
	LocalFileDirtyDirectory         LocalFileLockInspectionReason = "dirty_directory"
	LocalFileOwnerNotRegularFile    LocalFileLockInspectionReason = "owner_not_regular_file"
	LocalFileOwnerTooLarge          LocalFileLockInspectionReason = "owner_too_large"
	LocalFileOwnerInvalidUTF8       LocalFileLockInspectionReason = "owner_invalid_utf8"
	LocalFileOwnerIdentityChanged   LocalFileLockInspectionReason = "owner_identity_changed"
	LocalFilePermissionsWidened     LocalFileLockInspectionReason = "permissions_widened"
	LocalFileOwnerContractViolation LocalFileLockInspectionReason = "owner_contract_violation"
)

// LocalFileLockInspection is diagnostic evidence only; it never claims ownership.
type LocalFileLockInspection struct {
	State   LocalFileLockInspectionState
	Owner   string
	Reason  LocalFileLockInspectionReason
	Message string
}

// InspectLocalFileLock reports portable-lock shape without mutating it.
func InspectLocalFileLock(path string) (LocalFileLockInspection, error) {
	if err := validateLocalPath(path); err != nil {
		return LocalFileLockInspection{}, err
	}
	info, err := os.Lstat(path)
	if err != nil {
		if errors.Is(err, os.ErrNotExist) {
			return LocalFileLockInspection{State: LocalFileLockAbsent}, nil
		}
		return LocalFileLockInspection{}, localFileError(LocalFileIO, path, "inspect local lock path failed", err)
	}
	if !info.IsDir() || localFileInfoIsAlias(info) {
		return compromisedInspection(LocalFilePathNotDirectory, "lock path is not an unaliased directory"), nil
	}
	if err := validateLocalPOSIXPrivateMode(path, info, "lock directory", 0o022); err != nil {
		return compromisedInspection(LocalFilePermissionsWidened, err.Error()), nil
	}

	// Two names are enough to distinguish empty, pending/recovering transition,
	// exactly one published owner, and dirty state. Never enumerate an
	// arbitrarily large attacker-expanded directory.
	dir, err := os.Open(path)
	if err != nil {
		return LocalFileLockInspection{}, localFileError(LocalFileIO, path, "open local lock directory failed", err)
	}
	entries, readErr := dir.Readdirnames(2)
	closeErr := dir.Close()
	if readErr != nil && !errors.Is(readErr, io.EOF) {
		return LocalFileLockInspection{}, localFileError(LocalFileIO, path, "list local lock directory failed", readErr)
	}
	if closeErr != nil {
		return LocalFileLockInspection{}, localFileError(LocalFileIO, path, "close local lock directory failed", closeErr)
	}
	if len(entries) == 0 {
		return incompleteInspection("lock directory has no owner marker; acquisition, release, or recovery may have crashed mid-transition"), nil
	}
	if len(entries) == 1 && entries[0] == localFileOwnerPendingName {
		return incompleteInspection("owner publication is incomplete; pending owner marker is not ownership authority"), nil
	}
	if len(entries) == 1 && entries[0] == localFileOwnerRecoveringName {
		return incompleteInspection("owner recovery is in progress; recovery claim is not reusable ownership authority"), nil
	}
	if len(entries) != 1 || entries[0] != localFileOwnerName {
		return compromisedInspection(LocalFileDirtyDirectory, "lock directory must contain exactly one published owner marker"), nil
	}

	ownerPath := filepath.Join(path, localFileOwnerName)
	owner, err := readBoundedLocalFileLockOwner(path, ownerPath)
	if err != nil {
		var localErr *LocalFileLockError
		if errors.As(err, &localErr) && localErr.Kind == LocalFileCompromised && errors.Is(localErr.Cause, os.ErrNotExist) {
			return incompleteInspection("owner token disappeared during inspection"), nil
		}
		if errors.As(err, &localErr) && localErr.Kind == LocalFileCompromised {
			return compromisedInspection(reasonFromLocalOwnerError(localErr.Message), localErr.Message), nil
		}
		return LocalFileLockInspection{}, err
	}
	if len(owner) == 0 {
		return compromisedInspection(LocalFileOwnerContractViolation, "owner token is empty"), nil
	}
	if !utf8.Valid(owner) {
		return compromisedInspection(LocalFileOwnerInvalidUTF8, "owner token is not valid UTF-8"), nil
	}
	if utf8.RuneCount(owner) > localFileOwnerMaxCodepoints {
		return compromisedInspection(LocalFileOwnerContractViolation, "owner token exceeds the portable 512-code-point contract bound"), nil
	}
	return LocalFileLockInspection{State: LocalFileLockHeld, Owner: string(owner)}, nil
}

// readBoundedLocalFileLockOwner validates both the path identity and the
// already-open handle identity, then reads no more than the derived storage
// ceiling plus one byte. It is shared by inspection and normal release.
func readBoundedLocalFileLockOwner(lockPath, ownerPath string) ([]byte, error) {
	pathInfo, err := os.Lstat(ownerPath)
	if err != nil {
		if errors.Is(err, os.ErrNotExist) {
			return nil, localFileError(LocalFileCompromised, lockPath, "owner token is missing", err)
		}
		return nil, localFileError(LocalFileIO, lockPath, "inspect local lock owner token failed", err)
	}
	if !pathInfo.Mode().IsRegular() || localFileInfoIsAlias(pathInfo) {
		return nil, localFileError(LocalFileCompromised, lockPath, "owner token is not an unaliased regular file", nil)
	}
	if localFileHasMultipleLinks(pathInfo) {
		return nil, localFileError(LocalFileCompromised, lockPath, "owner token has multiple filesystem links; refusing aliased ownership state", nil)
	}
	if err := validateLocalPOSIXPrivateMode(lockPath, pathInfo, "owner token", 0o077); err != nil {
		return nil, err
	}
	if pathInfo.Size() > int64(localFileOwnerMaxUTF8Bytes) {
		return nil, localFileError(LocalFileCompromised, lockPath, "owner token exceeds the portable 2048-byte UTF-8 storage bound", nil)
	}

	ownerFile, err := os.Open(ownerPath)
	if err != nil {
		if errors.Is(err, os.ErrNotExist) {
			return nil, localFileError(LocalFileCompromised, lockPath, "owner token is missing", err)
		}
		return nil, localFileError(LocalFileIO, lockPath, "open local lock owner token failed", err)
	}
	defer ownerFile.Close()
	openedInfo, err := ownerFile.Stat()
	if err != nil {
		return nil, localFileError(LocalFileIO, lockPath, "stat opened local lock owner token failed", err)
	}
	if !openedInfo.Mode().IsRegular() || !os.SameFile(pathInfo, openedInfo) || localFileHasMultipleLinks(openedInfo) {
		return nil, localFileError(LocalFileCompromised, lockPath, "owner token identity changed or became multiply linked while opening; refusing raced path-to-handle state", nil)
	}
	if err := validateLocalPOSIXPrivateMode(lockPath, openedInfo, "opened owner token", 0o077); err != nil {
		return nil, err
	}
	if openedInfo.Size() > int64(localFileOwnerMaxUTF8Bytes) {
		return nil, localFileError(LocalFileCompromised, lockPath, "owner token exceeds the portable 2048-byte UTF-8 storage bound", nil)
	}

	owner, readErr := io.ReadAll(io.LimitReader(ownerFile, localFileOwnerMaxUTF8Bytes+1))
	if readErr != nil {
		return nil, localFileError(LocalFileIO, lockPath, "read local lock owner token failed", readErr)
	}
	if len(owner) > localFileOwnerMaxUTF8Bytes {
		return nil, localFileError(LocalFileCompromised, lockPath, "owner token exceeds the portable 2048-byte UTF-8 storage bound", nil)
	}
	if !utf8.Valid(owner) {
		return nil, localFileError(LocalFileCompromised, lockPath, "owner token is not valid UTF-8", nil)
	}
	return owner, nil
}

// RecoverLocalFileLock explicitly removes one clean portable lock after the
// operator independently confirms the former owner is inactive and local state
// is quiescent. An absent lock is an idempotent false/no-op result. Incomplete
// ownerless crash-window state is never auto-recovered because owner identity
// can no longer be authenticated.
func RecoverLocalFileLock(path, expectedOwner string, confirmedInactive bool) (bool, error) {
	if err := validateLocalPath(path); err != nil {
		return false, err
	}
	if !confirmedInactive {
		return false, localFileError(LocalFileInvalidInput, path, "explicit confirmed_inactive=true is required for recovery", nil)
	}
	if err := validateLocalOwner(path, expectedOwner); err != nil {
		return false, err
	}

	inspection, err := InspectLocalFileLock(path)
	if err != nil {
		return false, err
	}
	switch inspection.State {
	case LocalFileLockAbsent:
		return false, nil
	case LocalFileLockIncomplete:
		return false, localFileError(LocalFileCompromised, path, "incomplete lock state has no owner identity; refusing automatic recovery", nil)
	case LocalFileLockCompromised:
		return false, localFileError(LocalFileCompromised, path, inspection.Message, nil)
	}
	if inspection.Owner != expectedOwner {
		return false, localFileError(LocalFileCompromised, path, "owner token does not match expected recovery owner", nil)
	}

	finalInspection, err := InspectLocalFileLock(path)
	if err != nil {
		return false, err
	}
	if finalInspection.State != LocalFileLockHeld || finalInspection.Owner != expectedOwner {
		return false, localFileError(LocalFileCompromised, path, "local lock changed during recovery; refusing deletion", nil)
	}

	ownerPath := filepath.Join(path, localFileOwnerName)
	recoveringPath := filepath.Join(path, localFileOwnerRecoveringName)
	if err := os.Rename(ownerPath, recoveringPath); err != nil {
		if errors.Is(err, os.ErrNotExist) {
			after, inspectErr := InspectLocalFileLock(path)
			if inspectErr != nil {
				return false, inspectErr
			}
			if after.State == LocalFileLockAbsent || after.State == LocalFileLockIncomplete {
				return false, nil
			}
			return false, localFileError(LocalFileCompromised, path, "local lock changed while claiming recovery", err)
		}
		if errors.Is(err, os.ErrExist) {
			return false, localFileError(LocalFileCompromised, path, "recovery claim already exists", err)
		}
		return false, localFileError(LocalFileIO, path, "claim local lock recovery failed", err)
	}
	if err := os.Remove(recoveringPath); err != nil {
		return false, localFileError(LocalFileIO, path, "remove recovered owner claim failed", err)
	}
	if err := removeRecoveredLockDirectory(path); err != nil {
		return false, err
	}
	return true, nil
}

// Windows refuses to remove an otherwise empty directory while another
// concurrent inspector still has a directory handle open. The recovery claim
// has already removed the sole authoritative owner marker at this point, so a
// bounded retry is safe and lets the one destructive winner finish after those
// read-only handles close. POSIX keeps its single-attempt behavior.
func removeRecoveredLockDirectory(path string) error {
	attempts := 1
	if runtime.GOOS == "windows" {
		attempts = 50
	}
	var lastErr error
	for attempt := 0; attempt < attempts; attempt++ {
		if err := os.Remove(path); err == nil {
			return nil
		} else {
			lastErr = err
		}

		dir, openErr := os.Open(path)
		if openErr == nil {
			names, readErr := dir.Readdirnames(1)
			_ = dir.Close()
			if readErr == nil && len(names) > 0 {
				return localFileError(LocalFileCompromised, path, "remove recovered lock directory failed after recovery claim; directory is no longer empty", lastErr)
			}
		} else if errors.Is(openErr, os.ErrNotExist) {
			return nil
		}

		if runtime.GOOS != "windows" || attempt+1 == attempts {
			break
		}
		time.Sleep(2 * time.Millisecond)
	}
	return localFileError(LocalFileIO, path, "remove recovered lock directory failed after recovery claim", lastErr)
}

func incompleteInspection(message string) LocalFileLockInspection {
	return LocalFileLockInspection{
		State:   LocalFileLockIncomplete,
		Reason:  LocalFileOwnerMarkerMissing,
		Message: message,
	}
}

func compromisedInspection(reason LocalFileLockInspectionReason, message string) LocalFileLockInspection {
	return LocalFileLockInspection{State: LocalFileLockCompromised, Reason: reason, Message: message}
}

func reasonFromLocalOwnerError(message string) LocalFileLockInspectionReason {
	switch {
	case strings.Contains(message, "multiple filesystem links"), strings.Contains(message, "identity changed"):
		return LocalFileOwnerIdentityChanged
	case strings.Contains(message, "not an unaliased regular file"):
		return LocalFileOwnerNotRegularFile
	case strings.Contains(message, "2048-byte"):
		return LocalFileOwnerTooLarge
	case strings.Contains(message, "not valid UTF-8"):
		return LocalFileOwnerInvalidUTF8
	case strings.Contains(message, "permissions widened"):
		return LocalFilePermissionsWidened
	default:
		return LocalFileOwnerContractViolation
	}
}
